src/filter.rs

```diff
@@ -0,0 +1,297 @@
+    /// A sumgde filter if true, otherwise a clean filter.
+    pub to_worktree: bool,
+}
+
+impl FilterSource {
+    unsafe fn from_raw(raw: *const raw::git_filter_source) -> FilterSource {
+        let mode = raw::git_filter_source_mode(raw);
+        FilterSource {
+            to_worktree: (mode == raw::GIT_FILTER_TO_WORKTREE),
+            path: util::bytes2path(CStr::from_ptr(raw::git_filter_source_path(raw))
+                                   .to_bytes()).to_path_buf()
+        }
+    }
+}
+
+/// A trait representing a customized filter to be registered into libgit2
+/// filter chain.
+///
+/// Modeled after git2/sys/filter.h in a streaming manner.
+pub trait StreamFilter : Send + Sync + 'static {
+    /// Create a new stream to handle new blob, should return a initialized
+    /// WriteStream.
+    fn open_stream(&self, source: FilterSource, next: RawWriteStream)
+        -> Result<Box<WriteStream>, Error>;
+}
+
+unsafe impl Send for RawWriteStream {}
+
+impl WriteStream for RawWriteStream {
+    fn write(&mut self, buf: &[u8])
+        -> Result<(), Error> {
+        let code = unsafe {
+            ((*self.stream).write)((*self).stream,
+                buf.as_ptr() as *const c_char, buf.len())
+        };
+        call::try(code).and(Ok(()))
+    }
+
+    fn close(&mut self) -> Result<(), Error> {
+        let code = unsafe { ((*self.stream).close)(self.stream) };
+        call::try(code).and(Ok(()))
+    }
+}
+
+extern fn stream_write(stream: *mut git_writestream,
+                       buf: *const c_char,
+                       len: size_t) -> c_int {
+    let ws = stream as *mut Stream;
+    unsafe {
+        let input = slice::from_raw_parts(buf as *const u8, len);
+        match (*ws).handler.write(input) {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn stream_free(stream: *mut git_writestream) {
+    let _ = unsafe { Box::from_raw(stream as *mut Stream) };
+}
+
+extern fn stream_close(stream: *mut git_writestream) -> c_int {
+    let ws = stream as *mut Stream;
+    unsafe {
+        match (*ws).handler.close() {
+            Ok(..) => raw::GIT_OK,
+            Err(_) => raw::GIT_ERROR,
+        }
+    }
+}
+
+extern fn filter_init(_filter: *mut git_filter) -> c_int {
+    raw::GIT_OK
+}
+
+extern fn filter_shutdown(filter: *mut git_filter) {
+    let _ = unsafe { Box::from_raw(filter as *mut Filter) };
+    // dropped
+}
+
+extern fn filter_stream_init(out: *mut *mut raw::git_writestream,
+                        filter: *mut git_filter,
+                        _payload: *mut *mut c_void,
+                        src: *const raw::git_filter_source,
+                        next: *mut raw::git_writestream) -> c_int
+{
+    let filter = filter as *mut Filter;
+    let next = RawWriteStream { stream: next };
+    let source = unsafe { FilterSource::from_raw(src) };
+    let stream = unsafe { match (*filter).instance.open_stream(source, next) {
```

Comment:
Stylistically if there ends up being multiple unsafe blocks like this in a function I tend to just wrap them all up into one unsafe block.